### PR TITLE
Timeout for LockAcquireAction

### PR DESCRIPTION
**Pull request by akashdw**
Acquiring task lock for overlapping intervals causes a deadlock. Also the Overlord can run out of jetty threads when the TaskActionClient times out and retries while acquiring the same lock. This PR introduces a timeout for `LockAcquireAction`

### CODE REVIEW

**Comment by jihoonson** [comment]
What do you think about adding timeout for ```tryLock()``` as well? I think it will be useful.

> Acquiring task lock for overlapping intervals causes a deadlock.

Maybe we need to return ```TaskLockPosse``` when ```tryLock()``` or ```lock()``` is failed to figure out the current lock failure is due to deadlock or not?

**Comment by akashdw** [comment]
Thanks @jihoonson.
Timeout is not required for `tryLock()`  because it is non blocking and returns immediately.
Made changes to throw an exception and alert incase of overlapping interval.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java [hunk 1]
```diff
@@ -76,10 +78,21 @@
 
   @Inject
   public TaskLockbox(
-      TaskStorage taskStorage
+      TaskStorage taskStorage,
+      ServerConfig serverConfig
+  )
+  {
+    this.taskStorage = taskStorage;
+    this.lockTimeoutMillis = serverConfig.getMaxIdleTime().getMillis();
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: maxIdelTime looks to be deprecated (http://druid.io/docs/latest/configuration/indexing-service.html).

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java [hunk 2]
```diff
@@ -179,11 +192,20 @@ public int compare(Pair<Task, TaskLock> left, Pair<Task, TaskLock> right)
    */
   public TaskLock lock(final Task task, final Interval interval) throws InterruptedException
   {
+    long startTime = System.currentTimeMillis();
     giant.lock();
     try {
       Optional<TaskLock> taskLock;
       while (!(taskLock = tryLock(task, interval)).isPresent()) {
-        lockReleaseCondition.await();
+        lockReleaseCondition.await(lockTimeoutMillis, TimeUnit.MILLISECONDS);
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: ```lockTimeoutMillis``` should be updated if this line returns early.

File: extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 3]
```diff
@@ -1491,7 +1491,7 @@ private void makeToolboxFactory() throws IOException
         derby.metadataTablesConfigSupplier().get(),
         derbyConnector
     );
-    taskLockbox = new TaskLockbox(taskStorage);
+    taskLockbox = new TaskLockbox(taskStorage, 300000);
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java]: I think the timeout of 300 seconds is too long. Of course, waiting for 300 seconds is not usual, but if the lock is acquired after waiting for 300 seconds, it is likely to make unit tests failed eventually on travis due to the job timeout. I think it would be better to make tests failed earlier rather than waiting for such a long time. This is same for other tests.

File: indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java [hunk 4]
```diff
@@ -124,5 +152,42 @@ public void testTryLockAfterTaskComplete() throws InterruptedException
     Assert.assertFalse(lockbox.tryLock(task, new Interval("2015-01-01/2015-01-02")).isPresent());
   }
 
+  @Test(expected = InterruptedException.class)
```
Comment (akashdw) [inline indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java]: 👍

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 1 above)
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: serverConfig.getMaxIdleTime() corresponds to `druid.server.http.maxIdleTime` and is not deprecated.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java [hunk 5]
```diff
@@ -76,10 +78,21 @@
...
+      ServerConfig serverConfig
+  )
+  {
+    this.taskStorage = taskStorage;
+    this.lockTimeoutMillis = serverConfig.getMaxIdleTime().getMillis();
+  }
+
+  public TaskLockbox(
+      TaskStorage taskStorage,
+      long lockTimeoutMillis
```
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: `Should every lockAcquireAction have the same timeout?` Yes for now, it can also be extended where client can send a timeout.
For now `serverConfig.getMaxIdleTime() ` is the default timeout b/c even if task gets a lock after this period, overlord can not write the response in the closed socket.

File: extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java (hunk 3 above)
Comment (akashdw) [inline extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java]: 👍

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 2 above)
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: If this line returns early that means Task got the lock, not sure why final variable of this class should be updated ?

Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: A spurious wakeup can occur while awaiting (https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Condition.html). If it wakes up early but couldn't get a lock yet, the waiting time should be updated.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 5 above)
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: > For now serverConfig.getMaxIdleTime() is the default timeout b/c even if task gets a lock after this period, overlord can not write the response in the closed socket.

Using maxIdleTime as a default value sounds good. 

> Yes for now, it can also be extended where client can send a timeout.

I think we need to set different timeouts for each lock request in very near future because I'm working on prioritized locking (https://github.com/druid-io/druid/issues/4479, https://github.com/druid-io/druid/pull/1679) and this timeout feature will be great if tasks can set different timeouts according to their priorities.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 1 above)
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: Ah, right.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 5 above)
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: Sure, I'll create a separate PR where client can send locktimeout, if not default `serverConfig.getMaxIdleTime()` will be used.

Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: Ok. Thanks!

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 2 above)
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: 👍

**Comment by akashdw** [comment]
Thanks @jihoonson. addressed comments.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java [hunk 6]
```diff
@@ -274,11 +286,6 @@ public TaskLock lock(final Task task, final Interval interval) throws Interrupte
             log.makeAlert("Same Task is trying to acquire lock for overlapping interval")
                .addData("task", task.getId())
                .addData("interval", interval);
-            throw new ISE(
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: Why did you remove the code throwing an exception here? I think it would be better to throw an exception immediately if a deadlock is found rather than waiting for the lock request to be expired.

File: indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java [hunk 7]
```diff
@@ -54,7 +59,7 @@ public void setUp()
     EasyMock.expect(serverConfig.getMaxIdleTime()).andReturn(new Period(100));
     EasyMock.replay(serverConfig);
 
-    ServiceEmitter emitter = EasyMock.createMock(ServiceEmitter.class);
+    ServiceEmitter emitter  = EasyMock.createMock(ServiceEmitter.class);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java]: Please remove the unnecessary space.

File: indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java [hunk 8]
```diff
@@ -75,10 +80,11 @@ public void testLockForInactiveTask() throws InterruptedException
     lockbox.lock(NoopTask.create(), new Interval("2015-01-01/2015-01-02"));
   }
 
-  @Test(expected = IllegalStateException.class)
+  @Test
   public void testLockAfterTaskComplete() throws InterruptedException
   {
     Task task = NoopTask.create();
+    exception.expect(IllegalStateException.class);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java]: This should be ISE. Also, it would be better to check the exception message as well. Please refer to AppenderatorDriverFailTest as an example.

File: indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java [hunk 9]
```diff
@@ -124,43 +133,32 @@ public void testTrySmallerLock() throws InterruptedException
...
   {
     Assert.assertFalse(lockbox.tryLock(NoopTask.create(), new Interval("2015-01-01/2015-01-02")).isPresent());
   }
 
-  @Test(expected = IllegalStateException.class)
+  @Test
   public void testTryLockAfterTaskComplete() throws InterruptedException
   {
     Task task = NoopTask.create();
+    exception.expect(IllegalStateException.class);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/overlord/TaskLockboxTest.java]: This should be ISE. Also, it would be better to check the exception message as well. Please refer to AppenderatorDriverFailTest as an example.

File: indexing-service/src/test/java/io/druid/indexing/common/task/RealtimeIndexTaskTest.java [hunk 10]
```diff
@@ -951,7 +951,7 @@ private TaskToolbox makeToolbox(
   )
   {
     final TaskConfig taskConfig = new TaskConfig(directory.getPath(), null, null, 50000, null, false, null, null);
-    final TaskLockbox taskLockbox = new TaskLockbox(taskStorage);
+    final TaskLockbox taskLockbox = new TaskLockbox(taskStorage, 300000);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/common/task/RealtimeIndexTaskTest.java]: Would you reduce the timeout here as well?

File: indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java [hunk 11]
```diff
@@ -152,7 +152,7 @@
     }
     INDEX_MERGER.persist(index, persistDir, indexSpec);
 
-    final TaskLockbox tl = new TaskLockbox(ts);
+    final TaskLockbox tl = new TaskLockbox(ts, 300000);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactoryTest.java]: Would you reduce the timeout here as well?

File: indexing-service/src/test/java/io/druid/indexing/overlord/TaskLifecycleTest.java [hunk 12]
```diff
@@ -514,7 +514,7 @@ private TaskToolboxFactory setUpTaskToolboxFactory(
     Preconditions.checkNotNull(taskStorage);
     Preconditions.checkNotNull(emitter);
 
-    taskLockbox = new TaskLockbox(taskStorage);
+    taskLockbox = new TaskLockbox(taskStorage, 300000);
```
Comment (jihoonson) [inline indexing-service/src/test/java/io/druid/indexing/overlord/TaskLifecycleTest.java]: Would you reduce the timeout here as well?

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 6 above)
Comment (akashdw) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: @jihoonson Overlapping intervals causes deadlock for `lockAquireAction` and not for `lockTryAquireAction`. I think throwing exception for `tryLock`  is unnecessary. Other problem I could see is in the SegmentAllocationAction, https://github.com/druid-io/druid/blob/master/indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentAllocateAction.java?utf8=%E2%9C%93#L177. SegmentAllocateAction tries to acquire lock for the same task with different intervals and throwing exception for overlapping interval in `tryLock` might cause failure of segment allocations. 
For now I'm just alerting and logging which might be useful if someone run into deadlock issue.

File: indexing-service/src/test/java/io/druid/indexing/common/task/RealtimeIndexTaskTest.java (hunk 10 above)
Comment (akashdw) [inline indexing-service/src/test/java/io/druid/indexing/common/task/RealtimeIndexTaskTest.java]: Ahh.. Forget to reduce the timeout here. Done

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 6 above)
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: Sounds makes sense.

**Comment by jihoonson** [comment]
@akashdw thank you for the update. The latest patch looks good to me.

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java (hunk 5 above)
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java]: @akashdw I raised https://github.com/druid-io/druid/issues/4533.
